pyccel/codegen/printing/ccode.py

```diff
@@ -313,31 +335,54 @@ def copy_NumpyArray_Data(self, expr):
         Return
         ------
             String
-                Return a str that contains the declaration of a dummy data_buffer
-                       and a call to an operator which copies it to an NdArray struct
+                Return a str that contains the necessary declarations, allocations and copy operations
+                to create a NdArray
         """
+        self.add_import(c_imports['string'])
         rhs = expr.rhs
         lhs = expr.lhs
         if rhs.rank == 0:
             raise NotImplementedError(str(expr))
-        dummy_array_name = self.scope.get_new_name('array_dummy')
-        declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
-        dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
         arg = rhs.arg if isinstance(rhs, NumpyArray) else rhs
-
-        self.add_import(c_imports['string'])
         if isinstance(arg, Variable):
-            arg = self._print(arg)
-            cpy_data = "memcpy({0}.{2}, {1}.{2}, {0}.buffer_size);\n".format(lhs, arg, dtype)
-            return '%s' % (cpy_data)
-        else :
-            if arg.rank > 1:
-                # flattening the args to use them in C initialization.
-                arg = self._flatten_list(arg)
-            arg = ', '.join(self._print(i) for i in arg)
-            dummy_array = "%s %s[] = {%s};\n" % (declare_dtype, dummy_array_name, arg)
-            cpy_data = "memcpy({0}.{2}, {1}, {0}.buffer_size);\n".format(self._print(lhs), dummy_array_name, dtype)
-            return  '%s%s' % (dummy_array, cpy_data)
+            return f"array_copy_data(&{self._print(lhs)}, {self._print(arg)});\n"
+        order = lhs.order
+        declare_dtype = self.find_in_dtype_registry(self._print(rhs.dtype), rhs.precision)
+        dtype = self.find_in_ndarray_type_registry(self._print(rhs.dtype), rhs.precision)
```

Comment:
You could save `self._print(rhs.dtype)` to avoid calculating it twice

Reply:
https://github.com/pyccel/pyccel/pull/1128/commits/42a9d941db4c85babcde52931e811996876f431c